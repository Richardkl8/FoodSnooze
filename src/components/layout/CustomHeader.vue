<template>
    <header>
        <div class="container">
            <div class="center-items">
                <custom-h1 class="text-white">
                    <router-link class="text-white" to="/">
                        <b id="title">FoodSnooze</b>
                        <b id="title-mobile">FS</b>
                    </router-link>
                </custom-h1>
                <input
                    id="search"
                    placeholder="search for recipes"
                    v-model="query"
                    @keydown.enter="goToSearchPage(query)"/>
            </div>
        </div>
    </header>
</template>

<script>
    export default {
        name: 'CustomHeader',
        data() {
            return {
                query: '',
            };
        },
        watch: {
            $router() {
                this.query = '';
            },
        },
        methods: {
            goToSearchPage(query) {
                // Only executes when the query is different then the current query
                // and not on an empty query
                if (this.$route.params.query !== query && query) {
                    this.$router.push(`/search/${query}`);
                }
                this.query = '';
            },
        },
    };
</script>

<style lang="scss" scoped>
    header {
        background-color: #2DB396;
        width: 100vw;
        height: 10rem;
        display: flex;
        flex-direction: row;
        justify-content: center;
        align-items: center;
        color: #fff;
    }

    input {
        width: 40rem;
        padding: 1rem 2rem;
        font-size: 1.75rem;
        border: none;
        border-radius: 2rem;

        &:hover {
            cursor: text;
        }
    }

    .center-items {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
    }

    #title-mobile {
        display: none;
    }

    @media screen and (max-width: 768px) {
        #title {
            display: none;
        }

        #title-mobile {
            display: block;
            margin-right: 1rem;
        }
    }
</style>
