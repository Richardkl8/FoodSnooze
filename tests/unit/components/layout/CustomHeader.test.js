import { shallowMount } from '@vue/test-utils';
import CustomHeader from '../../../../src/components/layout/CustomHeader.vue';

const mocks = {
    $router: {
        push: jest.fn(),
    },
    $route: {
        params: {},
    },
};

describe('CustomHeader', () => {
    afterEach(() => {
        jest.resetAllMocks();
    });

    test('Should match snapshot', () => {
        const wrapper = shallowMount(CustomHeader, {
            stubs: [
                'custom-h1',
                'router-link',
            ],
        });
        expect(wrapper).toMatchSnapshot();
    });

    test('Should call \'goToSearchPage\' with query when user types in query and presses enter', () => {
        const wrapper = shallowMount(CustomHeader, {
            stubs: [
                'custom-h1',
                'router-link',
            ],
            mocks,
        });
        const spy = jest.spyOn(wrapper.vm, 'goToSearchPage');
        const searchBar = wrapper.find('input');

        searchBar.setValue('beef');
        searchBar.trigger('keydown.enter');

        expect(mocks.$router.push).toHaveBeenCalledWith('/search/beef');
        expect(spy).toHaveBeenCalledWith('beef');
    });

    test('Should reset search bar input when keydown.enter', () => {
        const wrapper = shallowMount(CustomHeader, {
            stubs: [
                'custom-h1',
                'router-link',
            ],
            mocks,
        });
        const searchBar = wrapper.find('input');

        searchBar.setValue('beef');
        expect(wrapper.vm.$data.query).toBe('beef');

        searchBar.trigger('keydown.enter');
        expect(wrapper.vm.$data.query).toBe('');
    });

    test('Should not go to page when search bar query is same as url query parameter', () => {
        const wrapper = shallowMount(CustomHeader, {
            stubs: [
                'custom-h1',
                'router-link',
            ],
            mocks: {
                ...mocks,
                $route: {
                    params: {
                        query: 'beef',
                    },
                },
            },
        });
        const searchBar = wrapper.find('input');

        searchBar.setValue('beef');
        searchBar.trigger('keydown.enter');

        expect(mocks.$router.push).not.toHaveBeenCalled();
    });
});
