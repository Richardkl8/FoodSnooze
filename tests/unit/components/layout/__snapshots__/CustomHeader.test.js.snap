// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomHeader Should match snapshot 1`] = `
<header>
  <div class="container">
    <div class="center">
      <custom-h1-stub class="text-white">
        <router-link-stub to="/" class="text-white"><b id="title">FoodSnooze</b> <b id="title-mobile">FS</b></router-link-stub>
      </custom-h1-stub> <input id="search" placeholder="search for recipes">
    </div>
  </div>
</header>
`;
